### PR TITLE
[memory_checker] Parameterize with combination of DuT and container.

### PR DESCRIPTION
**Pull request by yozhao101**
### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [x ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ x] 201911

#### What is the motivation for this PR?
This PR aims to parameterize this test case with different container names such that individual test case will be ran against each container. This method will avoid using a loop to iterate each container in a single test case. For testing on Dual ToR, this method can avoid the fixture function runs on upper ToR and test case runs on lower ToR. This PR will fix https://github.com/Azure/sonic-buildimage/issues/8430. 

#### How did you do it?
Currently we already has this infrastructure. The parameter `enum_dut_feature_container`  will be initialized with the combination of DuT and container names. The function `decode_dut_port_name(...)` will parse each string to get DuT name and container name.

#### How did you verify/test it?
I tested this change on Dual ToR `str2-7050cx3-acs-06, str2-7050cx3-acs-07` and single ToR `str-msn2700-03`.

#### Any platform specific information?
N/A

#### Supported testbed topology if it's a new test case?
N/A

### CODE REVIEW

File: tests/memory_checker/test_memory_checker.py [hunk 1]
```diff
@@ -36,6 +38,12 @@ def modify_monit_config_and_restart(duthost):
     Returns:
         None.
     """
+    dut_name, container_name = decode_dut_port_name(enum_dut_feature_container)
```
Comment (bingwang-ms) [inline tests/memory_checker/test_memory_checker.py:41]: ```decode_dut_port_name```? Is it a typo? ```decode_dut_port_name``` doesn't return dut_name and container name.

File: tests/memory_checker/test_memory_checker.py [hunk 2]
```diff
@@ -206,7 +214,12 @@ def test_memory_checker(duthosts, creds, enum_rand_one_per_hwsku_frontend_hostna
     Returns:
         None.
     """
-    duthost = duthosts[enum_rand_one_per_hwsku_frontend_hostname]
+    dut_name, container_name = decode_dut_port_name(enum_dut_feature_container)
```
Comment (bingwang-ms) [inline tests/memory_checker/test_memory_checker.py:217]: Same above.

File: tests/memory_checker/test_memory_checker.py (hunk 1 above)
Comment (bingwang-ms) [reply tests/memory_checker/test_memory_checker.py:41]: Oh, I see. It's extremely tricky. I don't suggest that usage although dutname|container_name has exactly same format as dutname|portname. Can you please add another interface like ```decode_dut_container_name```?

**Comment by bingwang-ms** [comment]
I checked the issue, and it feels like that the best way to address such issue is to ensure that the test is running on the DUT that ```stress``` utility was installed on. We have a fixture ```rand_selected_dut``` which will randomly select a DUT from both ToRs. Just use that fixture to replace ```duthost``` will address the issue, I guess.

File: tests/memory_checker/test_memory_checker.py (hunk 1 above)
Comment (yozhao101) [reply tests/memory_checker/test_memory_checker.py:41]: Good suggestion! Will update.

**Comment by yozhao101** [comment]
> I checked the issue, and it feels like that the best way to address such issue is to ensure that the test is running on the DUT that `stress` utility was installed on. We have a fixture `rand_selected_dut` which will randomly select a DUT from both ToRs. Just use that fixture to replace `duthost` will address the issue, I guess.

File: tests/memory_checker/test_memory_checker.py [hunk 3]
```diff
@@ -206,7 +214,12 @@ def test_memory_checker(duthosts, creds, enum_rand_one_per_hwsku_frontend_hostna
...
+    pytest_require(dut_name == enum_rand_one_per_hwsku_frontend_hostname and container_name == "telemetry",
+                   "Skips testing memory_checker of container '{}' on the DuT '{}' since another DuT '{}' was chosen."
+                   .format(container_name, dut_name, enum_rand_one_per_hwsku_frontend_hostname))
+    duthost = duthosts[dut_name]
```
Comment (yozhao101) [inline tests/memory_checker/test_memory_checker.py:221]: @bingwang-ms Regarding to the question we can use the `duthost = rand_selected_dut` to replace the line 221, I tested it did work. Also this test case currently is only running against `telemetry` container. 

However, there is an issue. Suppose we have Dual ToR `str2-7050cx3-acs-06` and `str2-7050cx3-acs-07`, The parameterized fixture `enum_dut_feature_container` includes a list of strings such as `str2-7050cx3-acs-06|lldp`, `str2-7050cx3-acs-06|database`, `str2-7050cx3-acs-06|snmp`, `str2-7050cx3-acs-06|telemetry`, ..., `str2-7050cx3-acs-07|lldp`, `str2-7050cx3-acs-07|database`, `str2-7050cx3-acs-07|snmp`, `str2-7050cx3-acs-07|telemetry`, ....

The function `decode_dut_port_name(...)` will parse each string to separate the DuT name from container name. If we use `duthost = rand_selected_dut` to replace the line 221 and `rand_selected_dut` is `str2-7050cx3-acs-06`, then this test case will run two times in `telemetry` container on DuT `str2-7050cx3-acs-06`. One time it is running for input string `str2-7050cx3-acs-06|telemetry` and another is for `str2-7050cx3-acs-07|telemetry`. 

Overall I think Ying's method using `pytest_require(dut_name == enum_rand_one_per_hwsku_frontend_hostname, ...)` does make sense.

Comment (bingwang-ms) [reply tests/memory_checker/test_memory_checker.py:221]: Why not add a filter to the output of ```decode_dut_port_name``` (actually ```decode_dut_feature```)? Checking  ```pytest_require(dut_name == enum_rand_one_per_hwsku_frontend_hostname, ...)``` will cause test randomly run or skip. I don't think it's expected.

File: tests/memory_checker/test_memory_checker.py (hunk 1 above)
Comment (yozhao101) [reply tests/memory_checker/test_memory_checker.py]: Updated!

**Review by bingwang-ms — APPROVED** [review]

**Pull request merged**